### DISCUSSION TITLE
/usr/bin occurs before /opt/homebrew/bin in your PATH --> reordering by macOS's /usr/libexec/path_helper

### DISCUSSION THREAD

**Discussion by piejanssens** [Homebrew/discussions#4747] [Getting started]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.7
ORIGIN: https://github.com/Homebrew/brew
HEAD: d4444b563e24ac7c05a93121c464c02dfa04d44f
Last commit: 18 hours ago
Core tap JSON: 28 Aug 10:15 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_BROWSER: open
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.42.0 => /opt/homebrew/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libDellMonitorSdkLib.dylib

Warning: /usr/bin occurs before /opt/homebrew/bin in your PATH.
This means that system-provided programs will be used instead of those
provided by Homebrew. Consider setting your PATH so that
/opt/homebrew/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/opt/homebrew/bin:$PATH"' >> ~/.zshrc

The following tools exist at both paths:
  git-receive-pack
  git-upload-archive
  git-shell
  git-upload-pack
  openssl
  git
```

### Description of issue

I'm updating PATH via `~/.zshenv` only. I have no `~/.profile`, no `~/.zprofile`.

From my `~/zshenv`:
````
case $(uname) in
Darwin)
  if [[ $(arch) == 'arm64' ]]; then
    eval $(/opt/homebrew/bin/brew shellenv)
  else
    eval $(/usr/local/bin/brew shellenv)
  fi
  export PATH=~/Library/Python/3.9/bin:$PATH
  export PATH="/Applications/Visual Studio Code.app/Contents/Resources/app/bin":$PATH
  export BROWSER='open'
  ;;
Linux)
  eval $(~/.linuxbrew/bin/brew shellenv)
  ;;
esac
````

PATH output:
```console
/Users/pietejanssens/.local/state/fnm_multishells/55831_1693229246589/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/usr/local/laps:/Library/Apple/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Users/pietejanssens/.local/state/fnm_multishells/55802_1693229246389/bin:/Applications/Visual Studio Code.app/Contents/Resources/app/bin:/Users/pietejanssens/Library/Python/3.9/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/opt/homebrew/opt/fzf/bin
```

I couldn't figure out why `/usr/bin` etc. were being inserted before other paths including Homebrew's.
Then I stumbled upon some info that confirms path_helper can change the order of the paths!
This affects macOS users who are using Homebrew + ZSH + setting up Hombrew paths in `.zshenv`.

This article explains the whole situation in great detail: https://gist.github.com/Linerre/f11ad4a6a934dcf01ee8415c9457e7b2

It would be good if `brew doctor` could detect this situation and instruct the user to execute `eval $(/opt/homebrew/bin/brew shellenv)` from `.zshrc` rather than doing so from `.zshenv.`

**Comment by osalbahr** [comment]
Can I ask why you’ve put the `eval` in `.zshenv` (as opposed to what the Homebrew installer recommends)?

**Reply by piejanssens** [reply]
I don't recall, but it certainly wasn't deliberate to not follow the instructors of the Homebrew installer script. I might have overlooked or changed up when reorganizing my dotfiles.

**Reply by osalbahr** [reply]
Ah, gotcha.

The main issue I can think of with implementing such `brew doctor` feature is that I don’t know of a trivial way for `brew` commands to detect whether the shell is actually `bash` or `fish` (as opposed to using the system’s shell). Even knowing the shell, it’s possible to put the `eval` in any file as long as the command gets to run at startup.

Also, the `eval` being in an appropriate startup file is not required by Homebrew as users can choose to manually run the `eval` command as needed (to avoid always messing with the environment variables).

**Reply by piejanssens** [reply]
Isn't it already detecting that since `brew doctor` was instructing me specificly to set a path in `~/.zshrc`?
```Consider setting your PATH so that
/opt/homebrew/bin occurs before /usr/bin. Here is a one-liner:
  echo 'export PATH="/opt/homebrew/bin:$PATH"' >> ~/.zshrc
```

Anyhow, I just thought opening this discussion might help someone with a similar issue in the future. 
I will close it as the issue itself is resolved. Thanks for chiming in!

**Reply by osalbahr** [reply]
I think `brew doctor` in that warning message is only directly checking `PATH` regardless of the shell startup files. You can try running the `eval` command manually and re-run `brew doctor` (the warning message should go away).

And yes I agree that opening this discussion is useful for others who might be facing the same issue.
